### PR TITLE
fix(request-response): cleanup connected on dial/listen failures

### LINKED ISSUE #4773: Failing debug_assert in Request-Response protocol
### Summary

I am hitting panics in debug mode at [this](https://github.com/libp2p/rust-libp2p/blob/master/protocols/request-response/src/lib.rs#L656) debug_assert. 

While the debug_assert is panic in request-response code, I do NOT think is a bug in request-response. I took a look at it and it seems to be related to a combination of factors.

* Multiple inflight dial requests must be happening
* Some of those inflight dials must get denied

I made a smallish reproducer by hacking the `file-sharing` example on this [branch](https://github.com/nathanielc/rust-libp2p/tree/repro/debug_assert). I added the `mdns` behaviour so that I can easily get multiple inflight dial requests. I added  the `libp2p_connection_limits` behaviour so that its easy to deny some of those requests. I set the max per peer limit to 2. 

Finally to reproduce I run these commands from the `example/file-sharing` directory:

Provider
```
RUST_LOG=info,libp2p_swarm=debug cargo run -- \
    --listen-addresses /ip4/0.0.0.0/tcp/40837 \
    --secret-key-seed 1 provide \
    --path README.md \
    --name README.md
```

Note using `/ip4/0.0.0.0/` binds to multiple listen addresses. I have 4-5 local interfaces that mean I listen on that many addresses. This means that when mdns discovers a peer it attempts to dial each address separately and we attempt to dial them as separate concurrent dial requests.

Retriever
```
RUST_LOG=info,libp2p_swarm=debug cargo run --  get --name README.md
```

It seems that the request-response behaviour is receiving an inconsistent set of events. It receives  both a call to `handle_established_outbound_connection` and a swarm event `DialFailure` for the same connection id. This breaks the assumptions of the request-response internal state management and triggers the debug_assert failure.

### Expected behavior

I expect debug_asserts to not panic. Additionally I expect that for a single connection id it is either considered failed or established but not both.

### Actual behavior

The debug_asserts panics because the request-response logic receives conflicting information about a connection.

### Relevant log output

```shell
Here are logs from the modified `file-sharing` example run.

Of note is that the `request-response` behaviour receives both a `handle_established_outbound_connection` event from the swarm as well as a `DialFailure` event from the swarm for the same connection id. This confuses the request-response internal state can causes the debug_assert failure.


[2023-10-31T22:09:57Z INFO  libp2p_request_response] handle_established_outbound_connection 1
[2023-10-31T22:09:57Z DEBUG libp2p_swarm] Connection established: PeerId("12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X") Dialer { address: "/ip4/192.168.0.136/tcp/40837/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X", role_override: Dialer }; Total (peer): 1.
[2023-10-31T22:09:57Z INFO  libp2p_request_response] handle_established_outbound_connection 3
[2023-10-31T22:09:57Z DEBUG libp2p_swarm] Connection established: PeerId("12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X") Dialer { address: "/ip4/192.168.12.189/tcp/40837/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X", role_override: Dialer }; Total (peer): 2.
[2023-10-31T22:09:57Z INFO  libp2p_request_response] handle_established_outbound_connection 4
[2023-10-31T22:09:57Z INFO  libp2p_request_response] on_dial_failure 4 Denied { cause: ConnectionDenied { inner: Exceeded { limit: 2, kind: EstablishedPerPeer } } }
[2023-10-31T22:09:57Z INFO  libp2p_request_response] handle_established_outbound_connection 2
[2023-10-31T22:09:57Z INFO  libp2p_request_response] on_dial_failure 2 Denied { cause: ConnectionDenied { inner: Exceeded { limit: 2, kind: EstablishedPerPeer } } }
[2023-10-31T22:10:17Z DEBUG libp2p_swarm] Connection closed with error KeepAliveTimeout: Connected { endpoint: Dialer { address: "/ip4/192.168.0.136/tcp/40837/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X", role_override: Dialer }, peer_id: PeerId("12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X") }; Total (peer): 1.
[2023-10-31T22:10:17Z INFO  libp2p_request_response] on_connection_closed 1 3 1
[2023-10-31T22:10:17Z DEBUG libp2p_swarm] Connection closed with error KeepAliveTimeout: Connected { endpoint: Dialer { address: "/ip4/192.168.12.189/tcp/40837/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X", role_override: Dialer }, peer_id: PeerId("12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X") }; Total (peer): 0.
[2023-10-31T22:10:17Z INFO  libp2p_request_response] on_connection_closed 3 2 0
thread 'async-std/runtime' panicked at /home/nathanielc/projects/rust-libp2p/protocols/request-response/src/lib.rs:640:9:
assertion `left == right` failed
  left: false
 right: true
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at examples/file-sharing/src/network.rs:125:24:
Sender not to be dropped.: Canceled
```

### Possible Solution

One solution is to make request-response more robust to these inconsistent events from the swarm but that seems like a fix that only hides the real problem.

### Version

v0.52.4

The reproducer branch is based off the `libp2p-v0.52.4` tag.

### Would you like to work on fixing this bug ?

Yes

### PR DESCRIPTION
**Pull request by nathanielc**
## Description

It is possible for dial and listen failures to be sent even after `handle_established_*_connection` methods are called on behaviours. Therefore we need to clean up any state about those failed connections.

Prior to this change the state would get out of sync and cause a debug_assert_eq panic.

Fixes: #4773.

## Notes & open questions

The tests do not pass locally yet. But I would like feedback on the approach before tracking down the failing test.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/request-response/src/lib.rs [hunk 1]
```diff
@@ -677,7 +677,24 @@ where
...
-    fn on_dial_failure(&mut self, DialFailure { peer_id, .. }: DialFailure) {
+    // Removes the connection for a peer if it exists.
+    fn remove_connection_for_peer(&mut self, peer: PeerId, connection_id: ConnectionId) {
+        if let Some(connections) = self.connected.get_mut(&peer) {
+            connections
+                .iter()
+                .position(|c| c.id == connection_id)
+                .map(|p: usize| connections.remove(p));
+        }
+    }
```
Comment (thomaseizinger) [inline protocols/request-response/src/lib.rs]: `ConnectionId`s are unique even across peers, meaning you should be able to remove a connection just by searching for the `ConnectionId`, regardless of the `PeerId`.

File: swarm/src/behaviour.rs [hunk 2]
```diff
@@ -488,6 +494,7 @@ pub struct ListenFailure<'a> {
     pub send_back_addr: &'a Multiaddr,
     pub error: &'a ListenError,
     pub connection_id: ConnectionId,
+    pub peer_id: Option<PeerId>,
```
Comment (thomaseizinger) [inline swarm/src/behaviour.rs]: I think this might be a useful addition in itself, however, I don't think it is needed for this particular case.

Mind posting this up as a separate PR?

**Review by thomaseizinger — COMMENTED** [review]
Thanks for tackling this!

Two comments :)

File: protocols/request-response/src/lib.rs [hunk 3]
```diff
@@ -677,7 +677,34 @@ where
         }
     }
 
-    fn on_dial_failure(&mut self, DialFailure { peer_id, .. }: DialFailure) {
+    // Removes the connection if it exists.
+    fn remove_connection(&mut self, connection_id: ConnectionId, peer: Option<PeerId>) {
+        if let Some(peer) = peer {
```
Comment (thomaseizinger) [inline protocols/request-response/src/lib.rs:682]: I don't understand the benefit of doing this. I'd much rather always remove by `ConnectionId` and simply ignore the `PeerId`.

Perhaps we should change our internal data structure to index by `ConnectionId` always?

File: protocols/request-response/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,9 @@
+## Unreleased
+
+- Remove failed connections from internal state. Fixes [Issue
+    4773](https://github.com/libp2p/rust-libp2p/issues/4773).
+    See [PR 4777](https://github.com/libp2p/rust-libp2p/pull/4777)
```
Comment (thomaseizinger) [inline protocols/request-response/CHANGELOG.md]:
```suggestion
- Correctly update internal state for failed connections.
  See [PR 4777](https://github.com/libp2p/rust-libp2p/pull/4777)
```

File: protocols/request-response/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,9 @@
+## Unreleased
```
Comment (thomaseizinger) [inline protocols/request-response/CHANGELOG.md]:
```suggestion
## 0.26.1 - unreleased
```

Comment (thomaseizinger) [reply protocols/request-response/CHANGELOG.md]: Please also bump the version in the manifest.

**Review by thomaseizinger — COMMENTED** [review]
Thank you! I've left some more comments.

File: protocols/request-response/CHANGELOG.md (hunk 5 above)
Comment (nathanielc) [reply protocols/request-response/CHANGELOG.md]: Can do I was curious how I was supposed to handle that bit.

File: protocols/request-response/src/lib.rs (hunk 3 above)
Comment (nathanielc) [reply protocols/request-response/src/lib.rs:682]: > Perhaps we should change our internal data structure to index by ConnectionId always?

Generally yes that makes sense. I took a look at the API the request response exposes and there are a few methods that want to look connections by peer id and do not have access to the connection id. They are 

* is_connected
* is_pending_outbound
* is_pending_inbound
* try_send_request

I am not too familiar with the access patterns of the behavior and not sure if the trade offs make sense. If we change the internal structure to be indexed by connection id these methods above would be O(n) instead of O(1) and removing a connection state for a failed connection would be O(1) vs O(n).

My hunch is that the above methods are a more common and more critical code path than removing connections for failed dials/listens, so therefore we should leave the internal index by PeerId.

Or I could change it to keep two indexes but that seems more complicated that we might want. 

Thoughts? Happy to go in either direction.

Comment (thomaseizinger) [reply protocols/request-response/src/lib.rs:682]: Lets go with O(n) for removing connections. I think there is an opportunity to build a data structure which can be used across all protocols for indexing some state `T` by both IDs but we don't have to build that in this PR.

Let me know if you are interested though :)

Comment (nathanielc) [reply protocols/request-response/src/lib.rs:682]: For this PR I agree lets keep it simple. However a shared data structure across proctocols sounds very useful

Comment (thomaseizinger) [reply protocols/request-response/src/lib.rs:682]: Do you mind opening an issue so we can track the idea of the shared data structure?

File: protocols/request-response/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,8 @@
+## 0.26.1 - unreleased
+
+- Correctly update internal state for failed connections.
+  See [PR 4777](https://github.com/libp2p/rust-libp2p/pull/4777)
```
Comment (thomaseizinger) [inline protocols/request-response/CHANGELOG.md:4]:
```suggestion
  See [PR 4777](https://github.com/libp2p/rust-libp2p/pull/4777).
```

File: protocols/request-response/Cargo.toml [hunk 7]
```diff
@@ -3,7 +3,7 @@ name = "libp2p-request-response"
 edition = "2021"
 rust-version = { workspace = true }
 description = "Generic Request/Response Protocols"
-version = "0.26.0"
+version = "0.26.1"
```
Comment (thomaseizinger) [inline protocols/request-response/Cargo.toml:6]: Missing this bump in the main `Cargo.toml`

File: protocols/request-response/src/lib.rs [hunk 8]
```diff
@@ -696,6 +723,17 @@ where
...
         }
+        // Its possible that this dial failure is for an existing connection.
+        // If so we need to remove the connection.
+        self.remove_connection(connection_id, peer_id);
+    }
+    fn on_listen_failure(&mut self, ListenFailure { connection_id, .. }: ListenFailure) {
+        // Its possible that this listen failure is for an existing connection.
+        // If so we need to remove the connection.
+        //
+        // TODO: Once  https://github.com/libp2p/rust-libp2p/pull/4818 merges we should pass in the
```
Comment (thomaseizinger) [inline protocols/request-response/src/lib.rs:734]: Stale comment. #4818 is unrelated now that we agreed to just always work based off `ConnectionId`.

File: protocols/request-response/src/lib.rs [hunk 9]
```diff
@@ -696,6 +723,17 @@ where
                 }
             }
         }
+        // Its possible that this dial failure is for an existing connection.
```
Comment (thomaseizinger) [inline protocols/request-response/src/lib.rs:726]: This comment is somewhat unnecessary because there is no such thing as a "non-existing" connection.

If we wanted to be precise we could explain that, depending on where this behaviour is in the tree, we could have already altered our state for this ID but another behaviour "denied" the connection after us.

Now that I am writing this, I had the idea that we can also fix this by only alterting the state in the `ConnectionEstablished` event.

Perhaps that might be the better solution after all?

Comment (thomaseizinger) [reply protocols/request-response/src/lib.rs:726]: Perhaps we should extend the docs of `NetworkBehaviour` to explain that.

cc @mxinden This can actually a bit of a problem that didn't occur to me before. Our strategy of "preloading" handlers can lead to us losing state if a `NetworkBehaviour` "after" us denies a connection.

To be safe, these "connection management" plugins should always come first in the behaviour tree. But also, we currently don't call `poll_close` on the handler when the connection is denied which means a handler doesn't have a way of giving the events back to the behaviour. But that is not an elegant solution anyway because it would mean _each_ protocol has to deal with thes lifecycle bugs.

I have to think about this some more but the better solution might be to change the `handle_established` callbacks to take `&self` and dispatch the `ConnectionEstablished` event only _after_ the handlers have been fully created.

That would avoid this partial altering of state at compile-time.

**Review by thomaseizinger — COMMENTED** [review]
I just had a realisation that this is a much bigger problem than I originally thought it is :(

File: protocols/request-response/src/lib.rs (hunk 9 above)
Comment (nathanielc) [reply protocols/request-response/src/lib.rs:726]: I don't know enough of the details of how preloading handlers works, however it was very confusing issue to debug that a connection that was "established" could have a listen or dial failure. I had wrongly assumed that once established a connection could fail but that would not be communicated as a listen/dial failure but rather just a connection error/closed message. 

So anything that makes that relationship clear would be good.

Comment (thomaseizinger) [reply protocols/request-response/src/lib.rs:726]: Yeah it didn't realise this issue until I thought deeper about this PR :(

It is unfortunately somewhat of a design flaw. Because `NetworkBehaviour`s are composed into a tree AND the `handle_` functions take `&mut self`, you might end up modifying our local state and a behaviour that is deeper down in the tree ends up rejecting the connection which essentially discards the data you just moved into the handler.

To avoid this, make sure any "connection management" behaviour like `libp2p-connection-limits` are listed first in your `Behaviour` that has `#[derive(NetworkBehaviour)]`.

Comment (nathanielc) [reply protocols/request-response/src/lib.rs:726]: > To avoid this, make sure any "connection management" behaviour like libp2p-connection-limits are listed first in your Behaviour that has #[derive(NetworkBehaviour)].

Thanks I'll make this change internally for now.

**Comment by thomaseizinger** [comment]
I am going to close this in favor of having to order the behaviours in a particular way (for now). Perhaps at a later point, we can fix the design issue and make those handle functions take `&self`.

**Comment by thomaseizinger** [comment]
I am wondering what is better:

- Keeping this as is, where the debug-assert makes people aware that there is inconsistent state
- Fixing this but silently losing state

**Comment by thomaseizinger** [comment]
I've recorded the issue in https://github.com/libp2p/rust-libp2p/issues/4870.

**Comment by nathanielc** [comment]
> I am wondering what is better:
> 
>     * Keeping this as is, where the debug-assert makes people aware that there is inconsistent state
> 
>     * Fixing this but silently losing state

In dev the debug assert failure breaks the process because it kills the running task and no further progress is made on the protocol. That's mildly annoying in dev because I have to restart the process. But its not a deal breaker and made me aware of the issue. 

In release mode it is just a memory leak. However the conditions that trigger this bug are in my experience much more common in dev than in a release mode. For example using mdns to get many concurrent dials. Its more common to use mdns in a dev setting than in a production setting. 

TL;DR leaving the debug_assert in place is fine.

**Comment by thomaseizinger** [comment]
Now that you have re-ordered the behaviours, you should not actually not hit this at all.
